### PR TITLE
Unify iterable action return type

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

Replaces the multiple return types of an iterable action with a single return type.

### Upgrade instructions

For an iterable action, the following changes should be applied:
- `#include <optional>`
- `#include "Parallel/AlgorithmExecution.hpp"`
- remove `#include "Parallel/AlgorithmMetafunctions.hpp"` if present
- Declare the return type of the `apply` function of an iterable action to be `Parallel::iterable_action_return_t`
- Replace the `return` statements of the iterable action in the following way:
   - If returning a one argument tuple of a DataBox, instead `return {Parallel::AlgorithmExecution::Continue, std::nullopt};`
   - If returning a two argument tuple of a DataBox and bool, instead `return {Parallel::AlgorithmExecution::Continue, std::nullopt};` if the bool is false and `return {Parallel::AlgorithmExecution::Pause, std::nullopt};` if the bool is true
   - If returning a two argument tuple of a DataBox and Parallel::AlgorithmExecution, instead return the Parallel::AlgorithmExecution follwed by `std::nullopt`
   - If returning a three argument tuple of a DataBox, Parallel::AlgorithmExecution, and size_t, just remove the first argument (i.e. the DataBox)

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/AlgorithmExecution.hpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  /// Stop the execution of iterable actions and do not allow their execution
+  /// until after a phase change. Simple actions will still execute.
+  Halt
+};
+
+/// Return type for iterable actions.  The std::optional can be used to specify
+/// the index of the action to be called next in the phase dependent action
+/// list.  Passing std::nullopt will execute the next action in the list.
+using iterable_action_return_t =
+    std::tuple<AlgorithmExecution, std::optional<size_t>>;
```
Comment (nilsvu) [inline src/Parallel/AlgorithmExecution.hpp:33]: What do you think about making this a struct so you can default the second parameter to `nullopt` in the constructor?

Comment (kidder) [reply src/Parallel/AlgorithmExecution.hpp:33]: I thought about it, but decided to go for making the return value explicit.

Comment (nilsvu) [reply src/Parallel/AlgorithmExecution.hpp:33]: Ok. My thought was that reading `{Continue, nullopt}` doesn't quite make clear what the second parameter means. Maybe there's another way to fix that?

File: docs/DevGuide/Parallelization.md [hunk 2]
```diff
@@ -509,65 +509,50 @@ current phase.  That is, it is equal to the `action_list` type alias
...
-returned (that is, the flow control flags are not modified).  For
-example, an action that stops the algorithm could return
+other chares using the `receive_data` function.
+
+Iterable actions can request that the algorithm be paused or halted
+for the current phase, and control which action in the current PDAL
+will be executed next. This is all done via the return value from the
+`apply` function.  The `apply` function for iterable actions must
+return a Parallel::iterable_action_return_t which is a
+`std::tuple<Parallel::AlgorithmExecution, std::optional<size_t>`. The
```
Comment (wthrowe) [inline docs/DevGuide/Parallelization.md]: Missing `>` for the end of the tuple.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Review by wthrowe — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**